### PR TITLE
Packaging and other project prep/cleanup

### PR DESCRIPTION
**Pull request by bording**
Includes relevant changes/commits from #1

This gets the packages being created much closer to being ready.

Still to do in followup PRs;
- Add real package descriptions
- Override default package tags of "nservicebus messaging". If all three packages end up having the same tags, the `PackageTags` property can be moved to Directory.Build.props
- Once the actual package reference dependencies are in place (from moving up to new ASP.NET Core/.NET Core) we should determine which (if any) should have version ranges.

### CODE REVIEW

**Comment by WojcikMike** [comment]
I've added package description, tags. For now I changed the reference to any ASPNET packages to be to major 1.* 

Any updates to the packages can be made in a separate PR's

**Comment by WojcikMike** [comment]
@HEskandari pls have a look

File: src/Particular.ViewModelComposition.Gateway/Particular.ViewModelComposition.Gateway.csproj [hunk 1]
```diff
@@ -4,17 +4,17 @@
...
   </PropertyGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\Particular.ViewModelComposition\Particular.ViewModelComposition.csproj" />
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore" Version="1.1.2" />
-    <PackageReference Include="Microsoft.AspNetCore.Routing" Version="1.1.2" />
+    <PackageReference Include="Microsoft.AspNetCore" Version="1.*" />
```
Comment (bording) [inline src/Particular.ViewModelComposition.Gateway/Particular.ViewModelComposition.Gateway.csproj]: We definitely should not be introducing wildcards. 👎 for this change.

File: src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj [hunk 2]
```diff
@@ -4,16 +4,16 @@
     <TargetFramework>netcoreapp1.1</TargetFramework>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\NServiceBus.snk</AssemblyOriginatorKeyFile>
-    <Description>TODO Add real package description</Description>
-    <!--<PackageTags>add package tags</PackageTags>-->
+    <Description>Package for view model composition in ASP.NET Core</Description>
+    <!--<PackageTags>aspnetcore ui composition viewmodel</PackageTags>-->
```
Comment (bording) [inline src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj]: Need to uncomment this if you want them to apply to the package.

**Comment by bording** [comment]
@WojcikMike I left some comments, but for me least, GitHub appears to be collapsing them.

Also, for the package tags, I think you should be aiming for more consistency there.

File: src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj [hunk 3]
```diff
@@ -4,16 +4,16 @@
     <TargetFramework>netcoreapp1.1</TargetFramework>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\NServiceBus.snk</AssemblyOriginatorKeyFile>
-    <Description>TODO Add real package description</Description>
-    <!--<PackageTags>add package tags</PackageTags>-->
+    <Description>Package for view model composition in ASP.NET Core</Description>
```
[the remark this answers is not in the document]
Comment (WojcikMike) [reply src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj]: Well, the gateway has: 
Package for view model composition in ASP.NET Core API
MVC has:
Package for view model composition in ASP.NET Core
And the base one has:
Package for view model composition

File: src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj (hunk 2 above)
Comment (WojcikMike) [reply src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj]: Uncommented

**Comment by WojcikMike** [comment]
@bording I addressed your comments. With regards to tags:
> Also, for the package tags, I think you should be aiming for more consistency there. 

I really don't know what to do with it. I tried to look for any composition UI frameworks to see what tags are used, and I struggled with it. I went with my best guess. Any ideas are more than welcome.

File: src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj (hunk 3 above)
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/Particular.ViewModelComposition.Mvc/Particular.ViewModelComposition.Mvc.csproj]: My thinking was, the platform is called ASPNET Core. The divisions are between MVC/WebAPI, but that's just me thinking. If you're happy with it, that's fine too.

**Comment by HEskandari** [comment]
I think we can ignore the compilation failure. The inspection is complaining about the old PR/commits which is now deleted.

**Comment by mikeminutillo** [comment]
Fixed :)

**Comment by bording** [comment]
> I really don't know what to do with it. I tried to look for any composition UI frameworks to see what tags are used, and I struggled with it. I went with my best guess. Any ideas are more than welcome.

Sorry, my comment wasn't as clear as it could have been. I meant internal consistency among the projects in this repo.

Here the current three PackageTags:
```
<PackageTags>aspnetcore api composition</PackageTags>
<PackageTags>aspnetcore ui composition viewmodel</PackageTags>
<PackageTags>ui composition viewmodel</PackageTags>
```
Why do only 2 out of 3 of them include `aspnetcore`? Why is `viewmodel` not in all 3? Isn't this all about ui composition? Why does the first one not have `ui`? What does `api' mean here?

I'm trying to understand why these all have differents tags. I would have expected them to be identical, or mostly identical.

**Comment by bording** [comment]
I have a similar question about the Descriptions:

```
<Description>Package for view model composition in ASP.NET Core API</Description>
<Description>Package for view model composition in ASP.NET Core</Description>
<Description>Package for view model composition.</Description>
```

Why do only 2 out of 3 of these mention ASP.NET Core? If these are distinct packages, I don't feel these descriptions explain what they are for.

What is the Gateway project? Why does it have a description about "API" ?


Also, the period at the end of the third one should be removed.

**Comment by WojcikMike** [comment]
@bording all good questions, let me pull in our Italian consultant to answer them :P 

@mauroservienti could you please help us write description and tags for the above packages?

**Comment by HEskandari** [comment]
The Gateway project provides extension methods for ASP.NET Core WebAPI. The MVC project provides similar extension methods for ASP.NET Core MVC. The other project doesn't even need a package, as the assembly should be embedded in the other two packages (rather than a package dependency). 

Here's what I propose:

```
<Description>Library for creating ViewModel composition in ASP.NET Core Web API</Description>
<Description>Library for creating ViewModel composition in ASP.NET MVC</Description>
```

**Comment by adamralph** [comment]
> The main library is what the other two depend on and it does not require a package.

I'm not sure about that.

> The main library is what the other two depend on...

This is exactly what packages are designed for. Embedding the assembly in both packages feels like an unnecessary hack. It could even end up causing some form of DLL-hell. Is there precedent for this?

**Comment by mauroservienti** [comment]
There are use cases that justify using both of them in the same app. I'm 👎 to shipping the base library in both packages. We should publish 3 packages, 1 for the engine, and the other 2 for MVC and for the basic Composition Gateway both dependant on the engine one.

**Comment by HEskandari** [comment]
@adamralph @mauroservienti My reasoning was that the heavy-lifting is done in the base package and I don't foresee releasing one without the other, but I guess that's possible.

Do you agree with this proposed description, each for the related project?

```
<Description>Library for creating ViewModel composition in ASP.NET Core Web API</Description>
<Description>Library for creating ViewModel composition in ASP.NET MVC</Description>
<Description>Library for creating ViewModel composition in ASP.NET Core applications</Description>
```

**Comment by bording** [comment]
@mauroservienti When consuming these packages, are you as a consumer interacting with APIs from all of the packages, or are the other two just infrastructure for the main package?

At this point I don't see anything in the proposed descriptions that hint toward that relationship at all.

Given the nature of transitive package dependencies, what package would be the top-level package?

@adamralph I'm not sure your proposal is still quite what we want. All of the projects reference ASP.NET Core packages, so they are are tied to ASP.NET Core.

**Comment by bording** [comment]
One other thing, I'm not quite sure what the distinction between "ASP.NET Core" and "ASP.NET Core MVC" is. There really aren't separate products here like there was for WebForms vs. MVC.

Perhaps all of this is hinting that we don't have the project/package names right yet?

**Comment by adamralph** [comment]
> Perhaps all of this is hinting that we don't have the project/package names right yet?

I tend to agree.

**Comment by mauroservienti** [comment]
So:

* `ViewModelComposition`: the composition engine that is unaware of the hosting technology, the only dependency here is ASP.Net Core
* `ViewModelComposition.Mvc`: the downstream that allows the composition engine to be hosted in a ASP.Net Core MVC application (depends on the engine)
* `ViewModelComposition.Gateway`: a package that allows to deploy a self-contained composition engine that behaves like a reverse-proxy/API Gateway (depends on the engine)

There is no point in using the engine alone, with no connection to MVC or a self-hosted gateway, in a similar way as there is no point in using NServiceBus Core without a transport.
The 3 packages could be collapsed in 1 single package, the only downside of which would be that there will be useless dependencies based on the hosting technology of choice.

**Comment by bording** [comment]
@mauroservienti So someone might use the Gateway package without the MVC package?

**Comment by mauroservienti** [comment]
@bording yes, like we do in the [workshop exercises](
https://github.com/Particular/Workshop/tree/master/exercises/01-composite-ui/after/Divergent.CompositionGateway) or one of the demos.

**Comment by bording** [comment]
So, it sounds like ViewModelComposition.Mvc and ViewModelComposition.Gateway are then top-level packages that someone would choose to install directly.

What about the following changes then:

ViewModelComposition -> ViewModelComposition.Engine
ViewModelComposition.Mvc -> ViewModelComposition

And keep ViewModelComposition.Gateway unchanged?

Also, looking at these names, I think "ViewModelComposition" might be too generic. Perhaps we need it to be "Particular.ViewModelComposition" instead?

**Comment by mauroservienti** [comment]
> ViewModelComposition -> ViewModelComposition.Engine

👍 

> ViewModelComposition.Mvc -> ViewModelComposition

On the fence, it gives me the impression that this is the important one, although the Gateway is probably the one we want to promote more since is not dependent on the view technology.

> Also, looking at these names, I think "ViewModelComposition" might be too generic. Perhaps we need it to be "Particular.ViewModelComposition" instead?

Namespaces and projects are already using the Particular prefix, aren’t packages also using it? I’m shortcutting in comments since I’m writing using the phone.

**Comment by bording** [comment]
> Namespaces and projects are already using the Particular prefix, aren’t packages also using it? I’m shortcutting in comments since I’m writing using the phone.

Oh yeah, I forgot about that! Nevermind then!

**Comment by bording** [comment]
> On the fence, it gives me the impression that this is the important one, although the Gateway is probably the one we want to promote more since is not dependent on the view technology.

Fair enough. I still don't have a good feel for how these packages are used.

Would it make more sense for ViewModelComposition.Gateway be named "ViewModelComposition" instead?

**Comment by adamralph** [comment]
I find that names like "Engine" or "Core" are usually a design smell. Packages should generally encapsulate something that can be named without having to reach for those words.

If the top level package is encapsulating view model composition for ASP.NET Core, that leads me to the name `ViewModelComposition.AspNetCore`.

The package for MVC would then be `ViewModelComposition.AspNetCore.Mvc`.

The "gateway" package is an interesting one. I understand that it can be used to build something that we are describing as a "gateway", but that sounds more like an _application_ of the package, rather than what the package actually provides.

_However_, the top level package actually depends on `Microsoft.AspNetCore.Routing`, which _does not_  depend on `Microsoft.AspNetCore` (I followed the dependency graph to make sure). Only the "gateway" project depends on `Microsoft.AspNetCore`, which then depends on all the hosting and server packages, etc. I wonder if that dependency is actually too high in the graph, and whether the "gateway" only actually needs to depend on a downstream package or packages.

BTW I'm assuming that the packages will all be prefixed with `Particular.` I'm omitting it for brevity.

**Comment by mauroservienti** [comment]
`Particular.ViewModelComposition.AspNetCore` and `Particular.ViewModelComposition.AspNetCore.Mvc` sounds good to me.

Let me try to explain what the gateway thing is in reality: it is an AspNetCore pipeline terminator, that handles all the incoming HTTP requests issued to that web server (it's not fully accurate but it should give you the idea). So we could say that it is a sort of standalone/self-sufficient view model composer, that doesn't require MVC to work properly. Now, you guys are the english native speakers, I'll you debate the proper name 😄

**Comment by adamralph** [comment]
Thanks @mauroservienti, that helps. Are there any other popular packages that do this kind of pipeline termination?

**Comment by mauroservienti** [comment]
Other then Nancy, or MVC I'm not aware of any. Probably [Ocelot](https://github.com/ThreeMammals/Ocelot) does a similar thing, but I'm not sure.
We could come up with a fantasy name, like Nancy is. What's wrong with `API Gateway`, or `Gateway`?

**Comment by adamralph** [comment]
Maybe we should get on a call to thrash this out. 😉

**Comment by mauroservienti** [comment]
Sure.

**Comment by mikeminutillo** [comment]
@mauroservienti @adamralph did the discussed call take place? Can we unblock this issue?

**Comment by HEskandari** [comment]
@mikeminutillo I have posted an update [here](https://github.com/Particular/BusinessValue/issues/30#issuecomment-418601438).

**Pull request merged**